src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -418,24 +432,31 @@ private ComponentState GetRequiredComponentState(int componentId)
 
         private ComponentState GetOptionalComponentState(int componentId)
             => _componentStateById.TryGetValue(componentId, out var componentState)
-                ? componentState
+                ? componentState 
                 : null;
 
-        private void ProcessRenderQueue()
+        /// <summary>
+        /// Processes the render queue.
+        /// </summary>
+        protected void ProcessRenderQueue()
         {
-            _isBatchInProgress = true;
-            var updateDisplayTask = Task.CompletedTask;
+            if (!CanRenderBatch)
+            {
+                return;
+            }
 
+            var updateDisplayTask = Task.CompletedTask;
+            _isBatchInProgress = true;
             try
             {
                 // Process render queue until empty
-                while (_batchBuilder.ComponentRenderQueue.Count > 0)
+                while (CanRenderBatch && RenderBatchBuilder.ComponentRenderQueue.Count > 0)
```

Comment:
Under what circumstances might `CanRenderBatch` become false when it was previously true during the same synchronous render cycle? I'm a bit unsure how to reason about whether new failure states now exist if we stop rendering a batch part-way through (we never did before).